### PR TITLE
Record failed plugins in Reserver Plugin

### PR DESCRIPTION
**Pull request by kerthcet**
#### What type of PR is this?
/kind feature
/sig scheduling

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

It's still useful to record failed plugins when running reserve plugins, e.g. in DRA plugin, when resource driver hasn't allocated the resource(inside `Reserve()`  function) yet, it will return unschedulable status.
https://github.com/kubernetes/kubernetes/blob/4c40d749006a8895f6718f2b624a3dbe975988ab/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go#L673-L691

Then when watching node update events for rescheduling, it will be more efficient by recording the unschedulable plugins, or we'll move pods to active or backoff queue anyway.

https://github.com/kubernetes/kubernetes/blob/4c40d749006a8895f6718f2b624a3dbe975988ab/pkg/scheduler/internal/queue/scheduling_queue.go#L771-L780

cc @Huang-Wei @pohly 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by kerthcet** [comment]
/hold 
in case of unexpected merge

**Review by SataQiu — COMMENTED** [review]
/lgtm

**Review by sanposhiho — CHANGES_REQUESTED** [review]
@kerthcet 

Thanks! It's helpful because this is what I planned to do as part of https://github.com/kubernetes/kubernetes/pull/117844. 
But, recording as `FailedPlugin` in state is not enough to tell the scheduling queue as unschedulableplugins. We only do that when `FitError`. Can you have a change around this as well so that the failed plugin recorded in reserve is handled in the scheduling queue? 
https://github.com/kubernetes/kubernetes/blob/c4f44b3de3a29ea430c67ff83cdaabbdefcac39a/pkg/scheduler/schedule_one.go#L871

**Review by Huang-Wei — COMMENTED** [review]
Fine with me treating the failure similar as other extension points (PreFilter/Filter/etc.). But I will leave it to @pohly to confirm.

A nit below.

**Comment by sanposhiho** [comment]
/triage accepted
/priority important-longterm

since it's part of https://github.com/kubernetes/kubernetes/pull/117844.

**Comment by sanposhiho** [comment]
/remove-lgtm

https://github.com/kubernetes/kubernetes/pull/118529#pullrequestreview-1468062052

**Comment by pohly** [comment]
> Fine with me treating the failure similar as other extension points (PreFilter/Filter/etc.). But I will leave it to @pohly to confirm.

Do we have this commit as part of a complete branch somewhere? Then I can test that the end result still works.

**Comment by sanposhiho** [comment]
I believe it won't work due to https://github.com/kubernetes/kubernetes/pull/118529#pullrequestreview-1468062052.
@kerthcet Could you check my comment? Not a super urgent, but it'll block DRA's EnqueueingHint stuff in near future.

**Comment by kerthcet** [comment]
I'm checking https://github.com/kubernetes/kubernetes/pull/118551 before going through this PR.

**Comment by kerthcet** [comment]
I think there's a bug about Permit plugin, so I'd like to fix it first.

**Comment by kerthcet** [comment]
/hold 
for https://github.com/kubernetes/kubernetes/pull/119046 merged first.

File: pkg/scheduler/framework/runtime/framework.go
Comment (kerthcet) [reply pkg/scheduler/framework/runtime/framework.go:1261]: I didn't find where we called it in the caller side, but we do when handling scheduling error, see https://github.com/kubernetes/kubernetes/blob/14bd9bc127eb8309d4c63770e848a07e96c8c30e/pkg/scheduler/schedule_one.go#L928-L930

they're with different log level and what we log in the PR containers more infos like the failed plugin name. So keep this the same with `RunPermitPlugin`? It has the same log.

**Comment by kerthcet** [comment]
This is quite similar to https://github.com/kubernetes/kubernetes/pull/119046 , but adding support for reserve plugin, would you mind to take another look @ahg-g

File: pkg/scheduler/framework/runtime/framework.go
Comment (ahg-g) [reply pkg/scheduler/framework/runtime/framework.go:1269]: That is fine, we do log the reason though, which should indicate something about the plugin. I guess the point is being consistent with what we do for Filters, but we can keep this at 4.

**Review by ahg-g — COMMENTED** [review]
/lgtm

**Pull request merged**
